### PR TITLE
[WIP] Schedule analytics

### PR DESCRIPTION
**Pull request by rooftopcellist**
##### SUMMARY

This adds a system job for scheduling the Automation Insights data collection so that users have more control over when it is run.  

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
4.0.0
```

##### ADDITIONAL INFORMATION
Outstanding work to be done:

> My original concern with that was that it shifts enablement(optin)/disablement(optout) up to the schedule of the system job instead of strictly on the settings. Meaning, if the user intended to disable analytics collection they would still see this job running which might concern them if it had a default schedule set at install time.

> Also, if they intended to enable this but we didn't set a default schedule then we would never get this and they would require two steps to actually enable it (opt-in, then add a schedule).

We can either link the INSIGHTS_TRACKING_STATE setting to enablement of the schedule for this system job, or we can make it so that this schedule is only enabled when you set INSIGHTS_TRACKING_STATE = True in the UI.  

It probably makes more sense to link them in the UI.

### CODE REVIEW

**Comment by rooftopcellist** [comment]
I attempted to make a functional test for this, but it seems it is not so simple to test the signals.py logic on the SQLite pytest db.  

```
@pytest.mark.django_db
def test_analytics_schedule_setting(get, patch, admin_user):
    from django.apps import apps
    from awx.main.models.jobs import SystemJobTemplate
    from awx.main.migrations._create_collection_system_jt import create_system_job_templates
    create_system_job_templates(apps, None)

    job_template = UnifiedJobTemplate.objects.filter(name='Automation Insights Collection').first()
    schedule = Schedule.objects.filter(name='Gather Automation Insights', unified_job_template=job_template.id).first()

    # disable insights tracking
    system_settings_url = reverse('api:setting_singleton_detail', kwargs={'category_slug': 'system'})
    patch(system_settings_url, user=admin_user, data={
        'INSIGHTS_TRACKING_STATE': False
    })
    
    url = reverse('api:system_job_template_schedules_list', kwargs={'pk': job_template.id})
    schedule_before = get(url, admin_user)
    
    assert schedule_before.data['results'][0]['enabled'] == False
    print(schedule_before.data['results'][0]['enabled'])
    
    # enable insights tracking
    patch(system_settings_url, user=admin_user, data={
        'INSIGHTS_TRACKING_STATE': True
    })
    
    schedule_after = get(url, admin_user)
    assert schedule_after.data['results'][0]['enabled'] == True
```

File: awx/conf/signals.py [hunk 1]
```diff
@@ -21,6 +22,14 @@ def handle_setting_change(key, for_delete=False):
     # When a setting changes or is deleted, remove its value from cache along
     # with any other settings that depend on it.
     setting_keys = [key]
+    # enabled/disable analytics sys job schedule with setting value
+    if key == 'INSIGHTS_TRACKING_STATE':
+        enabled = getattr(settings, key, None)
+        system_jt = SystemJobTemplate.objects.filter(name='Automation Insights Collection').first()
+        if system_jt is not None:
+            analytics_schedule = Schedule.objects.filter(name='Gather Automation Insights', unified_job_template=system_jt.id).first()
```
Comment (chrismeyersfsu) [inline awx/conf/signals.py]: If the schedule name changes, the enablement / disablement isn't going to work. I suggest getting all Schedule's related to the SystemJobTemplate and enable/disable via one Django `.update()` call.

File: awx/main/migrations/0073_v350_schedule_analytics_template.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    now_dt = now()
+    random_time = now() + timedelta(minutes=random.randint(-30,30))
+    random_schedule_time = random_time.strftime('%Y%m%dT%H%M%SZ')
+
+    sjt, created = SystemJobTemplate.objects.get_or_create(
+        job_type='gather_analytics --ship',
+        defaults=dict(
+            name='Automation Insights Collection',
+            description='Collect analytics data and send it to Automation Insights',
+            created=now_dt,
```
Comment (chrismeyersfsu) [inline awx/main/migrations/0073_v350_schedule_analytics_template.py]: Do we need to set the created or does it get auto-populated? Same with modified.

File: awx/main/migrations/0073_v350_schedule_analytics_template.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            modified=now_dt,
+            polymorphic_ctype=sjt_ct,
+        ),
+    )
+    if created:
+        sched = Schedule(
+            name='Gather Automation Insights',
+            rrule='DTSTART:%s RRULE:FREQ=DAILY;INTERVAL=1;COUNT=1' % random_schedule_time,
+            description='Automatically Generated Schedule',
```
Comment (chrismeyersfsu) [inline awx/main/migrations/0073_v350_schedule_analytics_template.py]: Make this better ... maybe `This schedule was generated by Tower during the install. Will periodically collect system statistical information and send it to Automation Insights. You can view this data via <url>`

File: awx/main/migrations/0073_v350_schedule_analytics_template.py (hunk 2 above)
Comment (rooftopcellist) [reply awx/main/migrations/0073_v350_schedule_analytics_template.py]: hmm, we specified it in previous system_job migrations.  but you are right, I don't see a reason why this wouldn't be autopopulated by the database based on the model.

File: awx/ui/client/src/management-jobs/card/card.controller.js [hunk 4]
```diff
@@ -48,66 +48,83 @@ export default
...
-                            resizable: false,
-                            onOpen: function(){
-                                $scope.$watch('prompt_for_days_form.$invalid', function(invalid) {
-                                    if (invalid === true) {
-                                        $('#prompt-for-days-launch').prop("disabled", true);
-                                    } else {
-                                        $('#prompt-for-days-launch').prop("disabled", false);
-                                    }
-                                });
+                        if(name === 'Automation Insights Collection') {
```
Comment (rooftopcellist) [inline awx/ui/client/src/management-jobs/card/card.controller.js:51]: This change removes the pop-up modal which is hard-coded for other system jobs.  It is unnecessary for this system job.

**Comment by rooftopcellist** [comment]
recheck

**Comment by rooftopcellist** [comment]
recheck

File: awx/conf/signals.py [hunk 5]
```diff
@@ -37,6 +38,13 @@ def handle_setting_change(key, for_delete=False):
             enter=not bool(for_delete),
         )
 
+    # enabled/disable analytics sys job schedule with setting value
```
Comment (ryanpetrello) [inline awx/conf/signals.py:41]: I understand the desire for backwards compatibility, but it seems a little odd to me that we'd have manual code to keep both a setting _and_ `SystemJobTemplate` schedule enablement in sync.  What happens if somebody goes and disables one/multiple schedules on this special `SystemJobTemplate`?  It seems like `settings.INSIGHTS_TRACKING_STATE` wouldn't/couldn't reflect that?

Maybe @matburt and @wenottingham feel differently, but I wonder if a better approach would be to just break compatibility and remove this old setting (and tell people going forward that the way you turn it on/off is by toggling on/off the pre-created schedule for this system job).

What I'm asking for here _would_ be a larger change and would mean we'd probably do something like this:

1.  In a new migration create this `SystemJobTemplate` and its default schedule and turn the schedule on/off based on the value of `settings. INSIGHTS_TRACKING_STATE` (so we can handle upgrades appropriately).
2. At license application time, turn it on/off _all_ schedules for this `SystemJobTemplate` based on whether or not you opt in.
3. Update docs to tell people how to turn it off, or manage _new_ schedules (and remove the path of updating it from the settings UI).

File: awx/main/migrations/_create_collection_system_jt.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    '''
+    Create default system job templates if not present. Create default schedules
+    only if new system job templates were created (i.e. new database).
+    '''
+
+    SystemJobTemplate = apps.get_model('main', 'SystemJobTemplate')
+    Schedule = apps.get_model('main', 'Schedule')
+    ContentType = apps.get_model('contenttypes', 'ContentType')
+    sjt_ct = ContentType.objects.get_for_model(SystemJobTemplate)
+    now_dt = now()
```
Comment (ryanpetrello) [inline awx/main/migrations/_create_collection_system_jt.py]: I'm fine w/ the `+-30 minute` before to stagger these, but we should probably aim _somewhere_ around 9pm-3am EST as a default.

I don't think having these run by default during US/European business hours is the best default.

Comment (rooftopcellist) [reply awx/main/migrations/_create_collection_system_jt.py]: :+1: I have changed this to set the start time for the schedule as midnight the night of the job creation.

File: awx/conf/signals.py (hunk 5 above)
Comment (rooftopcellist) [reply awx/conf/signals.py:41]: That sounds reasonable.  If @wenottingham agrees with this approach, I will make it so.

Comment (wenottingham) [reply awx/conf/signals.py:41]: Is it possible to just have the SystemJobTemplate noop/skip if the setting is off?

**Comment by rooftopcellist** [comment]
Closing this in favor of https://github.com/ansible/awx/pull/3895

File: awx/conf/signals.py (hunk 5 above)
Comment (rooftopcellist) [reply awx/conf/signals.py:41]: @wenottingham that would similarly require a signal.

Comment (wenottingham) [reply awx/conf/signals.py:41]: How would it require a signal for the management command to just check the setting and do nothing if not set?

Comment (rooftopcellist) [reply awx/conf/signals.py:41]: @wenottingham we already do that.  The issue with that is that it doesn't prevent the job from being run by the schedule, so the system job will periodically attempt to run, fail, and add noise to the logs.  But it is an option.

**Pull request closed without merging**